### PR TITLE
#110 Added SaaS support for the extension wrappers

### LINKED ISSUE #110: Add support for SaaS in the extension wrappers
Three things need to be updated:

- transformers_extension_wrapper.deploy_language_container should provide the new SaaS parameters to the TeLanguageContainerDeployer.create.
- sagemaker_extension_wrapper.deploy_language_container should provide the new SaaS parameters to the SmeLanguageContainerDeployer.create.
- extension_wrapper_common.encapsulate_bucketfs_credentials should encapsulate the SaaS parameters in a connection object.

### PR DESCRIPTION
**Pull request by ahsimb**
closes #110

### CODE REVIEW

File: exasol/nb_connector/extension_wrapper_common.py [hunk 1]
```diff
@@ -53,30 +81,45 @@ def encapsulate_bucketfs_credentials(
...
-    bfs_dest = (
-        f"{bfs_protocol}://{bfs_host}:{conf.get(CKey.bfs_port)}/"
-        f"{conf.get(CKey.bfs_bucket)}/{path_in_bucket};{conf.get(CKey.bfs_service)}"
-    )
-    # TLS certificate verification option shall be provided in the fragment field.
-    verify: Optional[bool] = (False if conf.get(CKey.trusted_ca)
-                              else optional_str_to_bool(conf.get(CKey.cert_vld)))
-    if verify is not None:
-        bfs_dest += f'#{verify}'
+    backend = (bfs.path.StorageBackend.saas if get_backend(conf) == StorageBackend.saas
```
Comment (ckunki) [inline exasol/nb_connector/extension_wrapper_common.py:87]: The code below only uses the property `name` of variable `backend`.

Would it be better to base the if statement on the root condition as returned by `get_backend(conf)` and use a local function to get the name?

```python
def bfs_backend(backend):
    if backend == StorageBackend.saas:
        return bfs.path.StorageBackend.saas
    else:
        return bfs.path.StorageBackend.onprem

backend = get_backend(conf)
if backend == StorageBackend.saas:
    ... bfs_backend(backend).name
else: 
    ... bfs_backend(backend).name
```

File: exasol/nb_connector/extension_wrapper_common.py [hunk 2]
```diff
@@ -53,30 +81,45 @@ def encapsulate_bucketfs_credentials(
...
+               else bfs.path.StorageBackend.onprem)
+    if backend == bfs.path.StorageBackend.onprem:
+        host = conf.get(CKey.bfs_host_name, conf.get(CKey.db_host_name))
+        protocol = "https" if str_to_bool(conf, CKey.bfs_encryption, True) else "http"
+        url = f"{protocol}://{host}:{conf.get(CKey.bfs_port)}"
+        verify: Optional[bool] = (False if conf.get(CKey.trusted_ca)
+                                  else optional_str_to_bool(conf.get(CKey.cert_vld)))
+        conn_to = (f'"backend":"{backend.name}", '
```
Comment (ckunki) [inline exasol/nb_connector/extension_wrapper_common.py:95]: I am wondering if a dict and `", ".join(....)` would make sense here.
Or even a kwarg function. 

```python
def conn_to(**kwargs):
    return ", ".join(f'"{k}":"{v}"' for k,v in kwargs.items())

url = "https://exasol.com"
service_name = "bfs"
bucket_name = "my_bucket"
path = "a/b/c"

c = conn_to(
    url=url,
    service_name=service_name,
    bucket_name=bucket_name,
    path=path,
)
```
yields 
```
"url":"https://exasol.com", "service_name":"bfs", "bucket_name":"my_bucket", "path":"a/b/c"
```

File: exasol/nb_connector/transformers_extension_wrapper.py [hunk 3]
```diff
@@ -210,17 +211,8 @@ def upload_model_from_cache(
...
-        conf.get(CKey.bfs_service), bfs_host,
-        int(str(conf.get(CKey.bfs_port))), str(conf.get(CKey.bfs_encryption)).lower() == 'true',
-        conf.get(CKey.bfs_user), conf.get(CKey.bfs_password), conf.get(CKey.bfs_bucket),
-        PATH_IN_BUCKET)
-
-    # Upload the downloaded model files into bucketfs
-    upload_path = get_model_path(conf.get(CKey.te_models_bfs_dir), model_name)
-    upload_model_files_to_bucketfs(cache_dir, upload_path, bucketfs_location)
+    raise NotImplementedError('(Uploading the model is temporarily unavailable. '
+                              'Awaiting changes in the Transformers Extension module.')
```
Comment (ckunki) [inline exasol/nb_connector/transformers_extension_wrapper.py:215]:
```suggestion
                              'Awaiting changes in the Transformers Extension module.)')
```

File: test/unit/test_extension_wrapper_common.py [hunk 4]
```diff
@@ -22,6 +25,24 @@ def filled_secrets(secrets) -> Secrets:
     return secrets
 
 
+@pytest.fixture
+def filled_saas_secrets(secrets) -> Secrets:
+    secrets.save(CKey.saas_url, "https://mock_saas.exasol.com")
+    secrets.save(CKey.saas_account_id, "w53lhsoifid794ms")
+    secrets.save(CKey.saas_database_name, "my_database")
+    secrets.save(CKey.saas_token, "xmfi58302lfj0ojf64ndk3ls")
```
Comment (ckunki) [inline test/unit/test_extension_wrapper_common.py:33]: Should this secret be committed to the git repo?

Comment (ckunki) [reply test/unit/test_extension_wrapper_common.py:33]: Or if these are mock secrets, maybe use a more obvious fake, like `token_123` and `account_123`.

**Review by ckunki — DISMISSED** [review]
Approved with some optional proposals.

File: exasol/nb_connector/extension_wrapper_common.py (hunk 1 above)
Comment (ahsimb) [reply exasol/nb_connector/extension_wrapper_common.py:87]: I don't think we need this `bfs_backend` function, but I take the point about the root condition.

File: exasol/nb_connector/extension_wrapper_common.py (hunk 2 above)
Comment (ahsimb) [reply exasol/nb_connector/extension_wrapper_common.py:95]: Very good idea, thanks.

File: exasol/nb_connector/transformers_extension_wrapper.py (hunk 3 above)
Comment (ahsimb) [reply exasol/nb_connector/transformers_extension_wrapper.py:215]: I messed it up a bit, thanks for spotting this.

File: test/unit/test_extension_wrapper_common.py (hunk 4 above)
Comment (ahsimb) [reply test/unit/test_extension_wrapper_common.py:33]: OK, will do.

**Review by ckunki — APPROVED** [review]

**Pull request merged**